x-pack/plugins/security/server/saved_objects/secure_saved_objects_client_wrapper.ts

```diff
@@ -173,25 +331,54 @@ export class SecureSavedObjectsClientWrapper implements SavedObjectsClientContra
     namespaces: string[],
     options: SavedObjectsDeleteFromNamespacesOptions = {}
   ) {
-    const args = { type, id, namespaces, options };
-    // To un-share an object, the user must have the "delete" permission in each of the target namespaces.
-    await this.ensureAuthorized(type, 'delete', namespaces, args, 'deleteFromNamespaces');
+    const auditor = await this.getScopedAuditor();
+    let response: {};
+    try {
+      // To un-share an object, the user must have the "delete" permission in each of the target namespaces.
+      const args = { type, id, namespaces, options };
+      await this.ensureAuthorized(type, 'delete', namespaces, args, 'deleteFromNamespaces');
 
-    return await this.baseClient.deleteFromNamespaces(type, id, namespaces, options);
+      response = await this.baseClient.deleteFromNamespaces(type, id, namespaces, options);
+    } catch (error) {
+      auditor.add(savedObjectUpdateEvent, {
+        action: 'saved_object_delete_from_namespaces',
+        objects: [{ type, id, namespaces }],
+        error,
+      });
+      throw error;
+    }
+
+    auditor.add(savedObjectUpdateEvent, {
+      action: 'saved_object_delete_from_namespaces',
+      objects: [{ type, id, namespaces }],
+    });
```

Comment:
These are all great points - Thanks for feedback. 

Not sure where to start ;) I think there will always be a level of "interpretation" with categorising all these different operations into a few predefined categories. I don't think it's necessarily an issue if they're not 100% technically correct as long as it makes sense from an auditors perspective. 

The mental model I used was that spaces is a bit like an attribute of an object and you can add space ids to it or remove them. We could instead also treat the operation as creating/deleting saved objects but I don't think that's correct / makes sense to someone who doesn't know the low level implementation details. The most important distinction is probably whether something was a read or write operation, not so much whether it was create or update? (might be wrong)

I didn't know about the delete case for orphaned saved objects, thanks for pointing that out. In that case that should probably be logged as an `{ action: 'deleteFromNamespaces', type: 'deletion' }`. Is there any way of knowing that that happened? Is it flagged in the UI that removing the object from that namespace will delete it?

Reply:
Another related point I'd like to discuss as part of this is that most of the feedback from the field was actually around auditors wanting to see exactly what was changed. (Down to field level diff/before/after) This pops up over and over again so is fairly high priority. We need to decide what will be logged at ElasticSearch vs. Kibana level but adding in some kind of a diff might help us deciding on a schema for events like `deleteFromNamespaces` / `addtoNamespaces`

Reply:
> I don't think it's necessarily an issue if they're not 100% technically correct as long as it makes sense from an auditors perspective.

Yeah, I agree.

> The most important distinction is probably whether something was a read or write operation, not so much whether it was create or update? (might be wrong but I've seen "Save" buttons in the UI that update a saved object actually using the `create` operation instead of `update`)

That's a good point. You can overwrite an existing object with a `create`, so in a way that's "updating" the object. But we really don't know whether an object was overwritten or not unless you make a pre-flight `get` operation.
I think leaving `create` as-is is fine.

> I didn't know about the delete case for orphaned saved objects, thanks for pointing that out. In that case that should probably be logged as an `{ action: 'deleteFromNamespaces', type: 'deletion' }`. Is there any way of knowing that that happened?

That makes sense. The `addToNamespaces` and `deleteFromNamespaces` operations currently only return empty objects, since the consumers that we had planned for didn't need any additional info. However, we can change these so they return the updated `namespaces` array to the consumer. I'll submit a PR for that and link it here.

> Is it flagged in the UI that removing the object from that namespace will delete it?

You can't do share/unshare objects in the UI (yet), this is part of Phase 1.5 of the Sharing Saved Objects project. However, the UI won't let you remove the current namespace from the object, so users will never run into this situation unless they are trying to use the API.

> We need to decide what will be logged at ElasticSearch vs. Kibana level but adding in some kind of a diff might help us deciding on a schema for events like `deleteFromNamespaces` / `addtoNamespaces`.

Perhaps if we include the args for each of these operations, and the resulting `namespaces` array, we cover what's needed? Alternatively we could have the saved objects client return the before/after.

Reply:
> I'll submit a PR for that and link it here.

PR is pending: #74995